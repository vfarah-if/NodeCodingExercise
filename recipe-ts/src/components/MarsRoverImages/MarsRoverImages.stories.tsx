import { Story, Meta } from '@storybook/react/types-6-0';
import React, { Fragment } from 'react';
import MarsRoverImages, { MarsRoverImagesProps } from './MarsRoverImages';

export default {
	title: 'Component/MarsRoverImages',
	component: MarsRoverImages,
} as Meta;

const MarsRoverImagesTemplate: Story<MarsRoverImagesProps> = (args) => (
	<Fragment>
		<MarsRoverImages {...args}/>
	</Fragment>
);

export const MarsRoverDefault = MarsRoverImagesTemplate.bind({});
MarsRoverDefault.args = {};

export const MarsRoverWithPresets = MarsRoverImagesTemplate.bind({});
MarsRoverWithPresets.args = { roverType: 'spirit', cameraType:'NAVCAM' };
