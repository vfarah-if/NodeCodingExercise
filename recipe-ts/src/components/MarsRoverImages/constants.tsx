export const ROVER_OPTIONS = [
	{
		value: 'curiosity',
		text: 'Curiosity',
	},
	{
		value: 'spirit',
		text: 'Spirit',
	},
	{
		value: 'opportunity',
		text: 'Opportunity',
	},
];

export const ROVER_CAMERA_OPTIONS = [
	{		
		value: '',
		text: 'Choose a camera',
	},
	{
		group: 'Spirit or Opportunity or Curiosity',
		value: 'NAVCAM',
		text: 'Navigation Camera',
	},
	{
		group: 'Spirit or Opportunity',
		value: 'PANCAM',
		text: 'Panoramic Camera',
	},
	{
		group: 'Curiosity',
		value: 'FHAZ',
		text: 'Front Hazard Avoidance Camera',
	},
	{
		group: 'Curiosity',
		value: 'RHAZ',
		text: 'Rear Hazard Avoidance Camera',
	},
	{
		group: 'Curiosity',
		value: 'MAST',
		text: 'Mast Camera',
	},
	{
		group: 'Curiosity',
		value: 'CHEMCAM',
		text: 'Chemistry and Camera Complex',
	}
];
