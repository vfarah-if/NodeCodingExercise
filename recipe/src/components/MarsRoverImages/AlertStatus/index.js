
import AlertStatus from './AlertStatus';
export default AlertStatus;    
