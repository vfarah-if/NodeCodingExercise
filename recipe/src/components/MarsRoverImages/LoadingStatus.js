import React from 'react';

export const LoadingStatus = ({ isLoading, displayValue }) => {
    return isLoading ? <p>{displayValue || 'Loading ...'}</p> : null;
}

export default LoadingStatus;
