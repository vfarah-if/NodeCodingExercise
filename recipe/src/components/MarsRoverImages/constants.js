export const ROVER_CAMERAS = {
	spirit: [
		{
			abbrev: 'NAVCAM',
			name: 'Navigation Camera',
		},
		{
			abbrev: 'PANCAM',
			name: 'Panoramic Camera',
		},
	],
	curiosity: [
		{
			abbrev: 'FHAZ',
			name: 'Front Hazard Avoidance Camera',
		},
		{
			abbrev: 'RHAZ',
			name: 'Rear Hazard Avoidance Camera',
		},
		{
			abbrev: 'MAST',
			name: 'Mast Camera',
		},
		{
			abbrev: 'CHEMCAM',
			name: 'Chemistry and Camera Complex',
		},
		{
			abbrev: 'NAVCAM',
			name: 'Navigation Camera',
		},
	],
	opportunity: [
		{
			abbrev: 'NAVCAM',
			name: 'Navigation Camera',
		},
		{
			abbrev: 'PANCAM',
			name: 'Panoramic Camera',
		},
	],
};