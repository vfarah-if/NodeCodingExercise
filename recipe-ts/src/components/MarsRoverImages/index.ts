
import MarsRoverImages from './MarsRoverImages';
export default MarsRoverImages;    
