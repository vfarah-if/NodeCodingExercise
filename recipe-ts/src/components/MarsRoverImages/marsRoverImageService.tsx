import config from '../../config';
import { NasaPhotos } from './MarsRoverImages';
const { NASAPIKey } = config;

export const marsRoverImageService = async (
	rover: string,
	camera: string
): Promise<NasaPhotos> => {
	const response = await fetch(
		`https://api.nasa.gov/mars-photos/api/v1/rovers/${rover}/photos?sol=100&camera=${camera}&api_key=${NASAPIKey}`
	);
	const result = (await response.json()) as NasaPhotos;
	return result;
};
