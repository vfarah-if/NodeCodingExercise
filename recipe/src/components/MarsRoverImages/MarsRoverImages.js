import React, { useState, useEffect } from 'react';

import LoadingStatus from '../LoadingStatus';
import AlertStatus from '../AlertStatus';
import ShinySelect from '../ShinySelect';
import Thumbnail from '../Thumbnail';
import Cards from '../Cards';
import Card from '../Card';
import config from '../../config';
import { ROVER_OPTIONS, ROVER_CAMERA_OPTIONS } from './constants';
import nasaLogo from '../../assets/nasa-logo-300x250.png';
import noImage from '../../assets/no-photo-400X300.png';
import './mars-rover-images.css';
const { NASAPIKey } = config;

export default function MarsRoverImages() {
	const [roverChosen, setRoverChosen] = useState('curiosity');
	const [cameraChosen, setCameraChosen] = useState('');
	const [error, setError] = useState(null);
	const [successMessage, setSuccessMessage] = useState(null);
	const [isLoading, setIsLoading] = useState(true);
	const [photos, setPhotos] = useState([]);
	const fetchMarsRoverPhotos = async () => {
		try {
			setIsLoading(true);
			if (roverChosen && cameraChosen && cameraChosen !== '') {
				const response = await fetch(
					`https://api.nasa.gov/mars-photos/api/v1/rovers/${roverChosen}/photos?sol=100&camera=${cameraChosen}&api_key=${NASAPIKey}`
				);
				const { photos } = await response.json();
				setPhotos(photos);
				setSuccessMessage(
					`Succeeded retrieving '${photos.length}' photos!`
				);
			} else {
				setPhotos([]);
			}
			return photos;
		} catch (error) {
			setError(error);
		} finally {
			setIsLoading(false);
		}
	};

	useEffect(() => {
		fetchMarsRoverPhotos();
		// eslint-disable-next-line
	}, [roverChosen, cameraChosen]);

	const handleRoverSelection = (event) => {
		setRoverChosen(event.target.value);
		setCameraChosen('');
		setSuccessMessage(null);
	};

	const handleCameraSelection = (event) => {
		setCameraChosen(event.target.value);
	};

	const hasNoPhotos = !photos?.length;

	return (
		<div className="Mars-Rover-Images">
			<div className="App-intro">
				<img
					src={nasaLogo}
					className="Mars-Rover-logo"
					alt="logo"
				></img>
			</div>
			<span>Data provided by NASA.</span>
			<br />
			<span>
				Each photo was taken on the rover's 100th sol (day) on Mars.
			</span>
			<br />
			<br />
			<ShinySelect
				displayName="Choose a Mars rover:"
				value={roverChosen}
				optionsList={ROVER_OPTIONS}
				onChange={handleRoverSelection}
			/>
			<br />
			<ShinySelect
				displayName="Choose a camera:"
				value={cameraChosen}
				optionsList={ROVER_CAMERA_OPTIONS}
				onChange={handleCameraSelection}
			/>
			<div>
				<LoadingStatus
					isLoading={isLoading}
					displayValue="Loading images ..."
				></LoadingStatus>
				<br />
				<AlertStatus alertType="success" message={successMessage} />
			</div>
			{hasNoPhotos ? (
				<div>
					<img alt="none found" src={noImage} />
				</div>
			) : (
				<Cards columnCount={5}>
					{photos?.map((item) => (
						<Card key={item.id}>
							<Thumbnail
								imageSource={item.img_src}
								description={`${item.id}`}
							/>
						</Card>
					))}
				</Cards>
			)}
			<AlertStatus alertType="error" message={error}></AlertStatus>
		</div>
	);
}
