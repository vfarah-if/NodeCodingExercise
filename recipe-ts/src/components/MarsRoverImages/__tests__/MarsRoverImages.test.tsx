import React from 'react';
import '@testing-library/jest-dom';
import { render, RenderResult, screen, waitFor } from '@testing-library/react';

import emptyResponse from './empty-response.json';
import curiosityFHAZResponse from './curiosity-fhaz-response.json';
import MarsRoverImages, { NasaPhotos } from '../MarsRoverImages';
import { marsRoverImageService } from '../marsRoverImageService';
jest.mock('../MarsRoverImageService');
const mockedService = marsRoverImageService as jest.MockedFunction<
	typeof marsRoverImageService
>;

describe('MarsRoverImages', () => {
	describe('When retrieving an empty response', () => {
		beforeEach(() => {
			mockedService.mockResolvedValue(emptyResponse as NasaPhotos);
		});

		test('should show success alert with zero images loaded', async () => {
			render(<MarsRoverImages roverType="spirit" cameraType="MAST" />);

			await waitFor(() => {
				expect(mockedService).toHaveBeenCalledTimes(1);
				// console.debug(screen.debug());
				expect(
					screen.getByText(`Succeeded retrieving '0' photos!`)
				).toBeInTheDocument();
			});
		});

		test('should show not found image', async () => {
			render(<MarsRoverImages roverType="spirit" cameraType="MAST" />);

			await waitFor(() => {
				expect(mockedService).toHaveBeenCalledTimes(1);
				// console.debug(screen.debug());
				expect(screen.getByAltText('none found')).toBeInTheDocument();
			});
		});
	});

	describe('When retrieving an full response', () => {
		beforeEach(() => {
			mockedService.mockResolvedValue(
				curiosityFHAZResponse as NasaPhotos
			);
		});

		test('should show success alert with zero images loaded', async () => {
			render(<MarsRoverImages roverType="curiosity" cameraType="FHA" />);

			await waitFor(() => {
				expect(mockedService).toHaveBeenCalledTimes(1);
				// console.debug(screen.debug());
				expect(
					screen.getByText(`Succeeded retrieving '4' photos!`)
				).toBeInTheDocument();
			});

			expect(
				screen.getByText(`Succeeded retrieving '4' photos!`)
			).toMatchSnapshot();
		});

		test('should show one cardslist and four cards', async () => {
			const { container } = render(
				<MarsRoverImages roverType="curiosity" cameraType="FHA" />
			);

			await waitFor(() => {
				expect(mockedService).toHaveBeenCalledTimes(1);
				expect(container.querySelectorAll('ul').length).toBe(1);
				expect(container.querySelectorAll('li').length).toBe(4);
			});

			// console.debug(screen.debug());
			expect(container.querySelector('ul')).toMatchSnapshot();
		});
	});

	//TODO: Figure this out later
	// describe('Select tests', () => {
	// 	let selectArray: NodeListOf<HTMLSelectElement>,
	// 		component: RenderResult<typeof import('@testing-library/dom/types/queries')>;

	// 	test('should have two selects', async () => {
	// 		await waitFor(() => {
	// 			component = render(
	// 				<MarsRoverImages roverType="curiosity" cameraType="FHA" />
	// 			);
	// 			selectArray = component.container.querySelectorAll('select');
	// 			expect(selectArray.length).toBe(2);
	// 		});
	// 		expect(selectArray[0]).toMatchSnapshot('Rovers');
	// 		expect(selectArray[1]).toMatchSnapshot('Cameras');
	// 	});
	// });
});
