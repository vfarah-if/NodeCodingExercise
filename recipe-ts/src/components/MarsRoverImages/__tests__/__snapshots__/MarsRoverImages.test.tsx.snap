// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarsRoverImages When retrieving an full response should show one cardslist and four cards 1`] = `
<ul
  style="display: grid; grid-template-columns: repeat(5, 1fr);"
>
  <li
    class="card"
  >
    <img
      alt="669"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FRA_406374643EDR_F0050178FHAZ00301M_.JPG"
    />
  </li>
  <li
    class="card"
  >
    <img
      alt="9487"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FLA_406377515EDR_F0050178FHAZ00302M_.JPG"
    />
  </li>
  <li
    class="card"
  >
    <img
      alt="9488"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FRA_406377515EDR_F0050178FHAZ00302M_.JPG"
    />
  </li>
  <li
    class="card"
  >
    <img
      alt="9489"
      class="thumbnail"
      src="http://mars.jpl.nasa.gov/msl-raw-images/proj/msl/redops/ods/surface/sol/00100/opgs/edr/fcam/FLA_406374643EDR_F0050178FHAZ00301M_.JPG"
    />
  </li>
</ul>
`;

exports[`MarsRoverImages When retrieving an full response should show success alert with zero images loaded 1`] = `
<div
  class="alert success"
>
  <span
    class="closebtn"
  >
    Ã—
  </span>
  Succeeded retrieving '4' photos!
</div>
`;
